### PR TITLE
[train][tune] v1 CheckpointManager only stores scoring metric if env var set

### PR DESCRIPTION
**Pull request by TimothySeah**
# Summary

When Ray Train v1 or Ray Tune users report metrics, we persist them in memory inside CheckpointManager, which can OOM. 

This change adds a new env variable, `TUNE_ONLY_STORE_CHECKPOINT_SCORE_ATTRIBUTE`, which makes it so CheckpointManager only saves the `checkpoint_score_attribute` metric in memory, reducing the memory footprint from O(checkpoints * metrics) to O(checkpoints).

Note that when `TUNE_ONLY_STORE_CHECKPOINT_SCORE_ATTRIBUTE` is set, we still persist all the metrics to disk, but when we construct the Tune [ResultGrid](https://github.com/ray-project/ray/blob/master/python/ray/tune/result_grid.py#L269), it only contains the `checkpoint_score_attribute`. This is just a temporary solution to unblock users; we can implement a more robust solution, such as reading metrics from disk in a streaming fashion, in the future. 

# Testing

I ran this driver script

```python
import ray
import ray.train
import ray.train.torch
import os

def train_func():
    dir_path = "/tmp/example_checkpoint_dir"
    os.makedirs(dir_path, exist_ok=True)
    ray.train.report(
        {"score": 2.5, "unsaved": 1.0}, 
        checkpoint=ray.train.Checkpoint.from_directory(dir_path))
    dir_path = "/tmp/another_example_checkpoint_dir"
    os.makedirs(dir_path, exist_ok=True)
    ray.train.report(
        {"unsaved": 0.5}, 
        checkpoint=ray.train.Checkpoint.from_directory(dir_path))
   
trainer = ray.train.torch.TorchTrainer(
    train_func,
    scaling_config=ray.train.ScalingConfig(num_workers=1),
    run_config=ray.train.RunConfig(
        checkpoint_config=ray.train.CheckpointConfig(
            num_to_keep=2,
            checkpoint_score_attribute="score",
            checkpoint_score_order="max",
        ),
    )
)
result = trainer.fit()
```

with this command `TUNE_ONLY_STORE_CHECKPOINT_SCORE_ATTRIBUTE=1 python ../playground/dummy_train_report.py`

I added print statements to CheckpointManager which verified that its internal state only has the score metric:

```
checkpoint_results: [TrainingResult(checkpoint=Checkpoint(filesystem=local, path=/Users/tseah/ray_results/TorchTrainer_2025-07-16_16-08-39/TorchTrainer_d0b6a_00000_0_2025-07-16_16-08-42/checkpoint_000000), metrics={'score': 2.5}), TrainingResult(checkpoint=Checkpoint(filesystem=local, path=/Users/tseah/ray_results/TorchTrainer_2025-07-16_16-08-39/TorchTrainer_d0b6a_00000_0_2025-07-16_16-08-42/checkpoint_000001), metrics={})]
```

I also inspected the `result.json` object to verify that it was still writing all the metrics to disk

```
(ray4) tseah@tseah-LV3607J62K TorchTrainer_d0b6a_00000_0_2025-07-16_16-08-42 % cat result.json
{"score": 2.5, "unsaved": 1.0, "timestamp": 1752707325, "checkpoint_dir_name": "checkpoint_000000", "should_checkpoint": true, "done": false, "training_iteration": 1, "trial_id": "d0b6a_00000", "date": "2025-07-16_16-08-45", "time_this_iter_s": 1.3460710048675537, "time_total_s": 1.3460710048675537, "pid": 67017, "hostname": "Mac.attlocal.net", "node_ip": "127.0.0.1", "config": {}, "time_since_restore": 1.3460710048675537, "iterations_since_restore": 1}
{"unsaved": 0.5, "timestamp": 1752707325, "checkpoint_dir_name": "checkpoint_000001", "should_checkpoint": true, "done": false, "training_iteration": 2, "trial_id": "d0b6a_00000", "date": "2025-07-16_16-08-45", "time_this_iter_s": 0.0030426979064941406, "time_total_s": 1.3491137027740479, "pid": 67017, "hostname": "Mac.attlocal.net", "node_ip": "127.0.0.1", "config": {}, "time_since_restore": 1.3491137027740479, "iterations_since_restore": 2}
```

I also tested this with tune/rllib in a workspace: 
* Without the env var set: internal CheckpointResults look like https://gist.github.com/TimothySeah/5a4d5ba9580e383b70195371520e4ece
* With the env var set, internal checkpoints look like
`checkpoint_results: [TrainingResult(checkpoint=Checkpoint(filesystem=local, path=/home/ray/ray_results/DQN_2025-07-16_18-31-09/DQN_CartPole-v1_b7aad_00000_0_2025-07-16_18-31-09/checkpoint_000004), metrics={})]`
and result.json looks like https://gist.github.com/TimothySeah/b2feb84c1300038b6748dc04472068b3

### CODE REVIEW

File: python/ray/train/_internal/checkpoint_manager.py [hunk 1]
```diff
@@ -90,17 +90,30 @@ def register_checkpoint(self, checkpoint_result: _TrainingResult):
...
+                ),
                 key=self._get_checkpoint_score,
             )
         else:
-            # If no metric is provided, just append (ordering by time of registration).
-            self._checkpoint_results.append(checkpoint_result)
+            # If no metric is provided, just append the checkpoint (ordering by time of registration).
+            self._checkpoint_results.append(
+                _TrainingResult(checkpoint=checkpoint_result.checkpoint, metrics={})
+            )
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/train/_internal/checkpoint_manager.py]: done.

File: python/ray/train/constants.py [hunk 2]
```diff
@@ -112,6 +112,16 @@ def _v2_migration_warnings_enabled() -> bool:
...
 RAY_TRAIN_ENABLE_STATE_TRACKING = "RAY_TRAIN_ENABLE_STATE_TRACKING"
 
+# Set this to 1 to only store the checkpoint score attribute with the Checkpoint
+# in the CheckpointManager. The Result will only have the checkpoint score attribute
+# but files written to disk like result.json will still have all the metrics.
+# Defaults to 0.
+# TODO: this is a temporary solution to avoid CheckpointManager OOM.
+# See https://github.com/ray-project/ray/pull/54642#issue-3234029360 for more details.
+RAY_TRAIN_ONLY_STORE_CHECKPOINT_SCORE_ATTRIBUTE = (
+    "RAY_TRAIN_ONLY_STORE_CHECKPOINT_SCORE_ATTRIBUTE"
```
Comment (matthewdeng) [inline python/ray/train/constants.py]: Think we should just call this TUNE since it'll primarily be for Tune after Ray Train V2 is fully enabled.

File: python/ray/train/_internal/checkpoint_manager.py [hunk 3]
```diff
@@ -99,6 +110,10 @@ def register_checkpoint(self, checkpoint_result: _TrainingResult):
                 key=self._get_checkpoint_score,
             )
         else:
+            if only_store_score_attr:
+                checkpoint_result = _TrainingResult(
+                    checkpoint=checkpoint_result.checkpoint, metrics={}
+                )
```
Comment (matthewdeng) [inline python/ray/train/_internal/checkpoint_manager.py]: Move the logic out of the if/else so that all the `only_store_score_attr` conditional logic is handled once in the beginning.

File: python/ray/train/tests/test_checkpoint_manager.py [hunk 4]
```diff
@@ -181,6 +181,72 @@ def test_nested_get_checkpoint_score(metrics):
     assert manager._get_checkpoint_score(tracked_checkpoint) == (True, 5.0)
 
 
+@pytest.mark.parametrize("has_score_attr", [True, False])
+def test_only_store_score_attr(has_score_attr, checkpoint_paths, monkeypatch):
+    monkeypatch.setenv("RAY_TRAIN_ONLY_STORE_CHECKPOINT_SCORE_ATTRIBUTE", "1")
+
+    # Set up CheckpointManager with 1 checkpoint.
```
Comment (matthewdeng) [inline python/ray/train/tests/test_checkpoint_manager.py]: What does 1 checkpoint mean here?

File: python/ray/train/tests/test_checkpoint_manager.py [hunk 5]
```diff
@@ -181,6 +181,72 @@ def test_nested_get_checkpoint_score(metrics):
...
+                )
+            ),
+            str(
+                _TrainingResult(
+                    checkpoint=Checkpoint.from_directory(checkpoint_paths[1]),
+                    metrics={},
+                )
+            ),
+            str(expected_tr3),
+        ]
```
Comment (matthewdeng) [inline python/ray/train/tests/test_checkpoint_manager.py]: Idk if it's a good idea to do string validation... probably better to just check the checkpoint/metrics attributes directly.

File: python/ray/train/constants.py (hunk 2 above)
Comment (TimothySeah) [reply python/ray/train/constants.py]: Done.

File: python/ray/train/_internal/checkpoint_manager.py (hunk 3 above)
Comment (TimothySeah) [reply python/ray/train/_internal/checkpoint_manager.py]: Done.

File: python/ray/train/tests/test_checkpoint_manager.py (hunk 4 above)
Comment (TimothySeah) [reply python/ray/train/tests/test_checkpoint_manager.py]: I meant that I wanted to
* Insert 1 checkpoint first
* Verify that we can insert 1 checkpoint before and after it.

But on second thought I think that's unnecessarily confusing and changed the comments.

File: python/ray/train/tests/test_checkpoint_manager.py (hunk 5 above)
Comment (TimothySeah) [reply python/ray/train/tests/test_checkpoint_manager.py]: Changed.

File: python/ray/train/constants.py [hunk 6]
```diff
@@ -112,6 +112,16 @@ def _v2_migration_warnings_enabled() -> bool:
 # Defaults to 0
 RAY_TRAIN_ENABLE_STATE_TRACKING = "RAY_TRAIN_ENABLE_STATE_TRACKING"
 
+# Set this to 1 to only store the checkpoint score attribute with the Checkpoint
+# in the CheckpointManager. The Result will only have the checkpoint score attribute
+# but files written to disk like result.json will still have all the metrics.
+# Defaults to 0.
+# TODO: this is a temporary solution to avoid CheckpointManager OOM.
+# See https://github.com/ray-project/ray/pull/54642#issue-3234029360 for more details.
+RAY_TUNE_ONLY_STORE_CHECKPOINT_SCORE_ATTRIBUTE = (
```
Comment (matthewdeng) [inline python/ray/train/constants.py]: Add this to the Tune docs https://docs.ray.io/en/master/tune/api/env.html

File: python/ray/train/tests/test_checkpoint_manager.py [hunk 7]
```diff
@@ -181,6 +182,57 @@ def test_nested_get_checkpoint_score(metrics):
...
+        )
+    )
+    assert len(manager.best_checkpoint_results) == 3
+    if has_score_attr:
+        assert manager.best_checkpoint_results[0].metrics == {"score": 1.0}
+        assert manager.best_checkpoint_results[0].checkpoint.path == checkpoint_paths[1]
+        assert manager.best_checkpoint_results[1].metrics == {"score": 3.0}
+        assert manager.best_checkpoint_results[1].checkpoint.path == checkpoint_paths[0]
+        assert manager.best_checkpoint_results[2].metrics == {}
+        assert manager.best_checkpoint_results[2].checkpoint.path == checkpoint_paths[2]
```
Comment (matthewdeng) [inline python/ray/train/tests/test_checkpoint_manager.py:226]: is this order correct? Should score==3 be at the top?

File: python/ray/train/tests/test_checkpoint_manager.py [hunk 8]
```diff
@@ -181,6 +182,57 @@ def test_nested_get_checkpoint_score(metrics):
...
+    else:
+        assert manager.best_checkpoint_results[0].metrics == {}
+        assert manager.best_checkpoint_results[0].checkpoint.path == checkpoint_paths[0]
+        assert manager.best_checkpoint_results[1].metrics == {}
+        assert manager.best_checkpoint_results[1].checkpoint.path == checkpoint_paths[1]
+        assert manager.best_checkpoint_results[2].metrics == {}
+        assert manager.best_checkpoint_results[2].checkpoint.path == checkpoint_paths[2]
```
Comment (matthewdeng) [inline python/ray/train/tests/test_checkpoint_manager.py:233]: is this order correct? Should it be insertion order ascending or descending?

File: python/ray/train/tests/test_checkpoint_manager.py (hunk 7 above)
Comment (TimothySeah) [reply python/ray/train/tests/test_checkpoint_manager.py:226]: I also thought that would be more intuitive but `best_checkpoint_result` does return the last element in the list: https://github.com/ray-project/ray/blob/master/python/ray/train/_internal/checkpoint_manager.py#L175. 

This means that TrainingResults that lack metrics will automatically be seen as "better." Maybe we should change that in the future but I don't want to introduce too many changes to this PR.

File: python/ray/train/tests/test_checkpoint_manager.py (hunk 8 above)
Comment (TimothySeah) [reply python/ray/train/tests/test_checkpoint_manager.py:233]: I agree flipping the order would be more intuitive but that's how it is right now: https://github.com/ray-project/ray/blob/master/python/ray/train/_internal/checkpoint_manager.py#L103

Comment (matthewdeng) [reply python/ray/train/tests/test_checkpoint_manager.py:233]: Ya sounds good I realized none of that logic was changed in this PR.

File: python/ray/train/constants.py (hunk 6 above)
Comment (TimothySeah) [reply python/ray/train/constants.py]: Added - not sure how many internal details to expose so ptal, thanks!

File: doc/source/tune/api/env.rst [hunk 9]
```diff
@@ -78,6 +78,10 @@ These are the environment variables Ray Tune currently considers:
   unsuccessful. After that, the trial is not restored to its previous checkpoint but rather from scratch.
   Default is ``0``. While this retry counter is taking effect, per trial failure number will not be incremented, which
   is compared against ``max_failures``.
+* **TUNE_ONLY_STORE_CHECKPOINT_SCORE_ATTRIBUTE**: If set to ``1``, will only store the ``checkpoint_score_attribute``
+  with each ``Checkpoint`` in memory. This means ``Result`` will only have the ``checkpoint_score_attribute``
+  but files written to disk like ``result.json`` will still have all the metrics.
+  Defaults to ``0``. This is a temporary solution to prevent the controller from OOM-ing.
```
Comment (matthewdeng) [inline doc/source/tune/api/env.rst]: Not sure if spacing is good, but something like this is more clear. No need to mention temporary solution or the disk files.

```suggestion
* **TUNE_ONLY_STORE_CHECKPOINT_SCORE_ATTRIBUTE**:  If set to `1`, only the metric defined by `checkpoint_score_attribute` will be stored with each `Checkpoint`. 
As a result, `Result.best_checkpoints` will contain only this metric, omitting others that would normally be included.  
This can significantly reduce memory usage, especially when many checkpoints are stored or when metrics are large.  
Defaults to `0` (i.e., all metrics are stored).
```

**Review by matthewdeng — APPROVED** [review]
